File: repository/repository.go
package repository

import (
	"fmt"
	"../domain"
	"../eventstore"
	"../eventbus"
)

type Repository struct {
	
}

func (repo Repository) Get(id string) domain.Task {
    fmt.Println("get " + id)
	
	task := domain.Task {}
	
    events := eventstore.GetInstance().LoadHistoryFor(id);

    for _, e := range events {
        task.Apply(e)
    }
	
	return task
}

func (repo Repository) Save(id string, e domain.Event, v int) {
    fmt.Println("save " + id)
	
    eventstore.GetInstance().SaveEvent(id, &e, v)
    eventbus.Publish(e)
}

File: CommandAndQueries/CommandAndQueries/Repository.swift
//
//  Repository.swift
//  CommandAndQueries
//
//  Created by Vitaly Baum on 11/10/14.
//  Copyright (c) 2014 Vitaly Baum. All rights reserved.
//

import Foundation

struct Repository {
    
    func get(id: String) -> Task {
        
        let task = Task()
        
        let events = EventStore.instance().loadHistoryFor(id);
        
        for e in events {
            task.apply(e)
        }
        
        return task
    }
    
    func save<T : Event>(id: String, e: T, v: Int) {
        
        EventStore.instance().save(id, e: e, v: v)
        EventBus.publish(e)
    }
}
